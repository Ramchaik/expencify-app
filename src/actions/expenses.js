import uuid from 'uuid';
import database from "../firebase/firebase";
/*
For Synchronous Redux Actions:

	To Do:
	 -- component calls the actions generator
	 -- action generator returns an object
	 -- component dispactches the object
	 -- redux stores changes


For Asynchronous Redux Actions:
	
	To Do:
	 -- component calls the actions generator
	 -- actions generator returns a function
	 -- component dispatches function
	 -- function runs (has the ability to dispatch other actions and do what ever it wants)
	 
*/

// ADD_EXPENSE
export const addExpense = (expense) => ({
	type: 'ADD_EXPENSE',
	expense
});

export const startAddExpense = (expenseData = {}) => {

	return (dispatch) => {
		const {
			description = '',
				note = '',
				amount = 0,
				createdAt = 0
		} = expenseData;

		const expense = {
			description,
			note,
			amount,
			createdAt
		};

		return database.ref("expenses")
			.push(expense)
			.then(ref => {
				dispatch(addExpense({
					id: ref.key,
					...expense
				}));
			})
	}
}

// Remove Expense
export const removeExpense = ({
	id = ''
}) => ({
	type: 'REMOVE_EXPENSE',
	id
});

export const startRemoveExpense = ({
	id
} = {}) => {
	return (dispacth) => {
		return database.ref(`expenses/${id}`)
			.remove()
			.then(ref => {
				dispacth(removeExpense({
					id
				}))
			});
	}
};

// EDIT EXPENSE
export const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
	type: "SET_EXPENSES",
	expenses
})

export const startSetExpenses = () => {
	return (dispatch) => {
		return database.ref("expenses")
			.once('value')
			.then((snapshot) => {
				const expenses = [];
				snapshot.forEach((childSnapshot) => {
					expenses.push({
						id: childSnapshot.key,
						...childSnapshot.val()
					});
				});
				dispatch(setExpenses(expenses));
			});
	}
}