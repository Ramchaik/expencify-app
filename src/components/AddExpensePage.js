import React from 'react';

const AddExpensePage = () => <div>This is a create expense page.</div>;

export default AddExpensePage;
