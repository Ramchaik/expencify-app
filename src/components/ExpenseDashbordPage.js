import React from 'react';
import ExpenseList from './ExpenseList';

const ExpenseDashbordPage = () => (
  <div>
    This is a ExpenseDashbordPage.
    <ExpenseList />
  </div>
);

export default ExpenseDashbordPage;
