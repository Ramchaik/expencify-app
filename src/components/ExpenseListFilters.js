import React from 'react';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calendarFocused => {
    this.setState(() => ({
      calendarFocused
    }));
  };

  onTextChange = e => this.props.setTextFilter(e.target.value);

  onSortChange = e => {
    const { value } = e.target;
    if (value === 'amount') {
      this.props.sortByAmount();
    } else {
      this.props.sortByDate();
    }
  };

  render() {
    const { props } = this;

    return (
      <div className='content-container'>
        <div className='input-group'>
          <div className='input-group__item'>
            <input
              type='text'
              className='text-input'
              value={props.filters.text}
              onChange={this.onTextChange}
              placeholder='Search expenses'
            />{' '}
          </div>
          <div className='input-group__item'>
            <select
              className='select'
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value='date'> Date </option>{' '}
              <option value='amount'> Amount </option>{' '}
            </select>{' '}
          </div>
          <div className='input-group__item'>
            <DateRangePicker
              startDate={props.filters.startDate}
              endDate={props.filters.endDate}
              onDatesChange={this.onDatesChange}
              showClearDates={true}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />{' '}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
