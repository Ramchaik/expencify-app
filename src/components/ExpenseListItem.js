import React from 'react';

const ExpenseListItem = ({ description, createdAt, amount }) => (
  <div>
    <h3>{description}</h3>
    <p>
      {amount} - {createdAt}
    </p>
  </div>
);

export default ExpenseListItem;
