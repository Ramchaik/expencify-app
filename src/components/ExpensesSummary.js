import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import totalExpenses from '../selectors/expenses-total';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => (
  <div>
    <h1>
      Viewing {expensesCount}
      {expensesCount === 1 ? ' expense' : ' expenses'} totalling{' '}
      {numeral(expensesTotal / 100).format('$0,0.00')}{' '}
    </h1>
  </div>
);

const mapStateToProps = state => ({
  expensesCount: selectExpenses(state.expenses, state.filters).length,
  expensesTotal: totalExpenses(selectExpenses(state.expenses, state.filters))
});

export default connect(mapStateToProps)(ExpensesSummary);
