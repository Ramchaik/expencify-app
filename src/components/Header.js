import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
  <header>
    <h1>Expencify</h1>
    <nav>
      <NavLink to='/' exact={true} activeClassName='is-active'>
        Home
      </NavLink>
      <NavLink to='/create' activeClassName='is-active'>
        Create
      </NavLink>
      <NavLink to='/edit' activeClassName='is-active'>
        Edit
      </NavLink>
      <NavLink to='/help' activeClassName='is-active'>
        Help
      </NavLink>
    </nav>
  </header>
);

export default Header;
