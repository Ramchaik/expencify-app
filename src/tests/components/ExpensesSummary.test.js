import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary with 1 expense', () => {
  const wrapper = shallow(
    <ExpensesSummary expensesCount={1} expensesTotal={expenses[0].amount} />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses', () => {
  const wrapper = shallow(
    <ExpensesSummary expensesCount={3} expensesTotal={158095} />
  );
  expect(wrapper).toMatchSnapshot();
});
