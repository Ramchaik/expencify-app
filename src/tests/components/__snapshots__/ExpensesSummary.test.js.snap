// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with 1 expense 1`] = `
<div>
  Viewing 
  1
   expense
   totalling 
  $1.95
   
</div>
`;

exports[`should render ExpensesSummary with multiple expenses 1`] = `
<div>
  Viewing 
  3
   expenses
   totalling 
  $1,580.95
   
</div>
`;
